File: benchmarks/solutions/ruby/r4-concurrent-fetch.rb
require "net/http"
require "uri"

urls = [
  "https://httpbin.org/get",
  "https://httpbin.org/status/404",
  "https://httpbin.org/delay/1",
  "https://httpbin.org/bytes/1024",
  "https://invalid.example.test",
]

semaphore = Queue.new
4.times { semaphore << true }

threads = urls.map do |url|
  Thread.new do
    semaphore.pop
    begin
      uri = URI.parse(url)
      resp = Net::HTTP.get_response(uri)
      puts "#{url}: #{resp.code} (#{resp.body.bytesize} bytes)"
    rescue => e
      puts "#{url}: error: #{e.message}"
    ensure
      semaphore << true
    end
  end
end

threads.each(&:join)

File: benchmarks/solutions/elixir/r4-concurrent-fetch.exs
urls = [
  "https://httpbin.org/get",
  "https://httpbin.org/status/404",
  "https://httpbin.org/delay/1",
  "https://httpbin.org/bytes/1024",
  "https://invalid.example.test",
]

urls
|> Task.async_stream_nolink(4, fn url ->
  case :httpc.request(:get, {String.to_charlist(url), []}, [timeout: 5000], body_format: :binary) do
    {:ok, {{_, status, _}, _, body}} ->
      "#{url}: #{status} (#{byte_size(body)} bytes)"
    {:error, reason} ->
      "#{url}: error: #{inspect(reason)}"
  end
end, timeout: 10_000)
|> Enum.each(fn
  {:ok, result} -> IO.puts(result)
  {:exit, reason} -> IO.puts("error: #{inspect(reason)}")
end)
